rfcs/0086-nixos-generation-apis.md

```diff
@@ -0,0 +1,195 @@
+- The name `nixos-rebuild` isn't strictly fitting to the `list-generations` parameter
+but a more general `nixos` command as proposed in
+[100578](https://github.com/NixOS/nixpkgs/pull/100578) and
+[54188](https://github.com/NixOS/nixpkgs/issues/54188).
+Though @edolstra [said it would be okay for now](https://github.com/NixOS/nixpkgs/issues/105910#issuecomment-754036275).
+
+
+## nixos-version
+
+Changing `nixos-version` so that a mistyped parameter throws an error improves
+error detection, but is a potentially breaking change, if there were scripts
+relying on mistyped flags. Though my opinion is, that it is better to throw
+meaningful errors instead of just doing something.
+
+# Alternatives
+[alternatives]: #alternatives
+
+## list-generations
+It would be possible to run `sudo nix-env -p /nix/var/nix/profiles/system --list-generations`,
+but that only prints the generation number and the build date, so a lot of
+invaluable information about the generations is missing. It requires a lot
+of knowledge about `nix-env`, profiles and the location/default profile for NixOS
+and (less important) privileged access.
+
+Putting the "(current)" tag into it's own column would make differentiating it
+from the configuration revision easier. On the other hand it could lead to a
+less readable output (because of separation) like
+```
+Generation  Build-date               NixOS version           Kernel   Configuration Revision
+1           Sat 2021-02-05 12:25:55  21.03.20210206.9bd6a6c  5.9.9    2199dd2822638f7dce2a9895930c51abd1723234
+2           Sat 2021-02-06 23:10:44  21.03.20210206.e6a8458  5.9.9
+3           Sat 2021-02-06 23:19:24  21.03.20210206.e6a8458  5.9.9
+4           Sat 2021-02-06 23:20:21  21.03.20210206.e6a8458  5.9.9                                              (current)
+5           Sat 2021-02-06 23:29:12  21.03.20210206.e6a8458  5.9.9
+```
+
+Currently it only lists the generation of the `$profile` with the `system` on as default.
+It would be possible to list all available generations, but that would result in
+a much more difficult to understand output, since the profiles would need to be
+labled. For a better discovery of available profiles, a `list-profiles` option
+could be implemented, but I guess this is out of scope for this RFC.
+
+
+# Unresolved questions
+[unresolved]: #unresolved-questions
+
+## list-generations
+
+- Is there any other useful information to show?
+- Default value for configuration revision field if unset? (probably not)
+- Truncate revision (maybe if hash only)?
+- Put the `(current)` tag in it's own column?
+- Show only system-profile by default?
+- How to handle machine-readability?
+- Provide stable API (e.g. `--porcelain`) for futureprooving output changes for
+scripts? (probably unnecessary, since usage in scripts probably limited)
+- Use `column` format for human readability? => Evaluation is eager
+- Use `pager` to avoid spamming terminal?
+- Is the order fine?
```

Comment:
All of these are good questions. The best way to answer them is to play with an implementation to get a feel for it.

Targetting the interface for human-consumption first will allow more flexibility and to change the output as needed. Then when it's stable, add a --json output for machine consumption.